### PR TITLE
crimson/os/seastore: avoid onode/omap laddr hint conflicts as much as possible

### PR DESCRIPTION
**Pull request by xxhdx1985126**
When profiling crimson-osd + seastore, we found that `BtreeLBAManager::alloc_extent` occupied much of cpu time. It seems that this is because there are many conflicts when trying to alloc omap onodes.

The following is the cpu time overhead of `BtreeLBAManager::alloc_extent` before and after this pr's optimization(purple blocks correspond to `BtreeLBAManager::alloc_extent`):
Before:
image

After:
image

Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/os/seastore/omap_manager/btree/btree_omap_manager.h
Comment (athanatos) [inline src/crimson/os/seastore/omap_manager/btree/btree_omap_manager.h:31]: Is the idea here to avoid having the omap addresses for a particular onode cluster around the hint?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/omap_manager/btree/btree_omap_manager.h:31]: yes:-)

**Comment by cyx1231st** [comment]
I think we should be able to observe a much smaller value from metrics `LBA::alloc_extents_iter_nexts / LBA::alloc_extents` after this change.

File: src/crimson/os/seastore/seastore.cc
Comment (athanatos) [inline src/crimson/os/seastore/seastore.cc:1084]: Why not modify get_hint() to generate a random hint within the onode's range?  That would remove the need for the omap specific logic.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/seastore.cc:1084]: I think object data writes also rely on `Onode::get_hint()`, if we randomize `Onode::get_hint()`, it might lead to laddr conflicts between object data extents and omap extents, am I right?

Comment (athanatos) [reply src/crimson/os/seastore/seastore.cc:1084]: Objects tend not to use both the data payload and omap payload heavily, generally one or the other.  Also, if we're randomizing the addresses within that space, even if both are in use, shouldn't it still be ok?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/seastore.cc:1084]: > Objects tend not to use both the data payload and omap payload heavily, generally one or the other. 

Yes, I totally agree.

> Also, if we're randomizing the addresses within that space, even if both are in use, shouldn't it still be ok?

It seems that the current object data handler tries to reserve objects' whole laddr space at the time of the first write. And I think we probably will want to set objects' size to 'GB' level when crimson is used in production environments, because that will reduce the number of objects and give us the possibility to cache all objects' metadata in memory, which would dramatically increase system performance as no metadata related I/O is needed for data reads/writes. So, say we set objects' size to 1 GB and reserve its whold laddr space at the time of the first data write, if we randomize `Onode::get_hint()`, later omap extent alloc would have a high chance to collide with that 1 GB data laddr space reservation. Am I right?:-)

Comment (athanatos) [reply src/crimson/os/seastore/seastore.cc:1084]: object_data_handler only reserves 16MB.  I don't want to worry about GB size objects at this time -- there are a lot of other things we'd need to change for that to be a good idea.

Comment (athanatos) [reply src/crimson/os/seastore/seastore.cc:1084]: I think simply randomizing the hint will get us the same performance advantage at this time with much less complexity.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/seastore.cc:1084]: will do:-)

Comment (xxhdx1985126) [reply src/crimson/os/seastore/seastore.cc]: done:-)

File: src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h
Comment (athanatos) [inline src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]:
```
crush & (((uint64_t)0x1 << (64 - ONODE_LADDR_SPACE_ORDER)) - 1)
```
This seems to be masking a 32 bit value with ((1<<32) - 1), 0xFFFFFFFF?

I think we can make this even simpler, just modify get_lba_hint here to return a uniform random value:

```
+constexpr uint32_t ONODE_HINT_RANGE = 32<<20; // Randomize hints over 32MB
 static laddr_t get_lba_hint(shard_t shard, pool_t pool, crush_hash_t crush) {
   // FIXME: It is possible that PGs from different pools share the same prefix
   // if the mask 0xFF is not long enough, result in unexpected transaction
   // conflicts.
-  return ((uint64_t)(shard & 0XFF)<<56 |
-          (uint64_t)(pool  & 0xFF)<<48 |
-          (uint64_t)(crush       )<<16);
+  auto hint = ((uint64_t)(shard & 0XFF)<<56 |
+              (uint64_t)(pool  & 0xFF)<<48 |
+              (uint64_t)(crush       )<<16);
+  // Randomize hints to avoid addresses for a particular object clustering
+  // at the hint.
+  return hint + (std::rand() % ONODE_HINT_RANGE);
 }
```

Comment (xxhdx1985126) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: > This seems to be masking a 32 bit value with ((1<<32) - 1), 0xFFFFFFFF?

Oh, sorry, this is a mistake, should be 1<<32

> I think we can make this even simpler, just modify get_lba_hint here to return a uniform random value:
> 
> ```
> +constexpr uint32_t ONODE_HINT_RANGE = 32<<20; // Randomize hints over 32MB
>  static laddr_t get_lba_hint(shard_t shard, pool_t pool, crush_hash_t crush) {
>    // FIXME: It is possible that PGs from different pools share the same prefix
>    // if the mask 0xFF is not long enough, result in unexpected transaction
>    // conflicts.
> -  return ((uint64_t)(shard & 0XFF)<<56 |
> -          (uint64_t)(pool  & 0xFF)<<48 |
> -          (uint64_t)(crush       )<<16);
> +  auto hint = ((uint64_t)(shard & 0XFF)<<56 |
> +              (uint64_t)(pool  & 0xFF)<<48 |
> +              (uint64_t)(crush       )<<16);
> +  // Randomize hints to avoid addresses for a particular object clustering
> +  // at the hint.
> +  return hint + (std::rand() % ONODE_HINT_RANGE);
>  }
> ```

I think `crush << 16` leaves only 64KB between onodes, so I used `(crush & (1 << 32)) << 32` to enlarge that interval to 4GB, so onodes are unlikely to collide with each other.

Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: > I think crush << 16 leaves only 64KB between onodes

`crush << 16` randomizes onode locations at the 48-bit range addresses because `crush` is a hashed value.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: > I think we can make this even simpler, just modify get_lba_hint here to return a uniform random value:
> 
> ```
> +constexpr uint32_t ONODE_HINT_RANGE = 32<<20; // Randomize hints over 32MB
>  static laddr_t get_lba_hint(shard_t shard, pool_t pool, crush_hash_t crush) {
>    // FIXME: It is possible that PGs from different pools share the same prefix
>    // if the mask 0xFF is not long enough, result in unexpected transaction
>    // conflicts.
> -  return ((uint64_t)(shard & 0XFF)<<56 |
> -          (uint64_t)(pool  & 0xFF)<<48 |
> -          (uint64_t)(crush       )<<16);
> +  auto hint = ((uint64_t)(shard & 0XFF)<<56 |
> +              (uint64_t)(pool  & 0xFF)<<48 |
> +              (uint64_t)(crush       )<<16);
> +  // Randomize hints to avoid addresses for a particular object clustering
> +  // at the hint.
> +  return hint + (std::rand() % ONODE_HINT_RANGE);
>  }
> ```

I don't quite follow, do you mean that we randomize onode data laddr start point and omap laddr hints within the 32MB space?

I think this would mean that half of omap extents laddr hints will collide with onode data laddrs, because onode data is 16MB long, occupying half of the 32MB space. Am I right?

Comment (athanatos) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: PG log objects don't have data payloads, and so do not have a 16MB reservation.

Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: Probably `constexpr uint32_t ONODE_HINT_RANGE = 1<<30 ; //Randomize hints over 1GB` for now?

This also means collision over 2^18 onodes per pool, should be enough?

Comment (athanatos) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: This doesn't actually work, the omap implementation caches a single hint value.  Needs a bit more of a refactor.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: > PG log objects don't have data payloads, and so do not have a 16MB reservation.

ok, will do:-)

Comment (athanatos) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: Objects don't typically have 32MB of data.  Randomizing over 1GB would tend to cause blocks from adjacent objects to overlap to a larger degree.  We could make it a config and test different values.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: > Huh, I guess omap_root_t::hint doesn't actually get persisted as part of omap_root_le_t, it gets set fresh each time we read it. This should work fine.

It seems that omap_root_t::hint will stay the same until the next `BtreeOMapManager::initialize_omap` is called again, am I right? If so, omap extents' hint will still have a high possibility of collision, am I right?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: > Objects don't typically have 32MB of data. Randomizing over 1GB would tend to cause blocks from adjacent objects to overlap to a larger degree. We could make it a config and test different values.

will do:-)

Comment (athanatos) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: https://github.com/athanatos/ceph/commit/5ec5bd54d5f3c692cf09ce7e5939477ab568b390 is I think a bit cleaner.  It updates the hint mechanism in Onode to provide different data and metadata hints and updates the callers appropriately.

Comment (athanatos) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: > It seems that omap_root_t::hint will stay the same until the next `BtreeOMapManager::initialize_omap` is called again, am I right? If so, omap extents' hint will still have a high possibility of collision, am I right? We still have to further randomize omap extent laddr hints even if omap_root_t::hint doesn't change, right?

It's confusing, but no.  omap_root_le_t doesn't actually store the hint.  omap_root_le_t::get(laddr_t hint) populates the returned omap_root_t with the passed hint each time it's invoked.  Really, hint shouldn't be part of omap_root_t.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h:51]: > [athanatos@5ec5bd5](https://github.com/athanatos/ceph/commit/5ec5bd54d5f3c692cf09ce7e5939477ab568b390) is I think a bit cleaner. It updates the hint mechanism in Onode to provide different data and metadata hints and updates the callers appropriately. It does ensure that metadata and data hints won't overlap as well in case we end up with objects that use both.

Ah, finally, I see:-) Will cherry-pick that commit, and repush the pr:-)

**Comment by xxhdx1985126** [comment]
@athanatos @cyx1231st I've cherry-picked that commit, and made the space reservation configurable, please take a look again.

**Comment by tchaikov** [comment]
might want to take a look. this does not compile.

**Comment by xxhdx1985126** [comment]
> might want to take a look. this does not compile.

Sorry, I didn't build tests in my environment, will take care of it:-)

File: src/crimson/os/seastore/object_data_handler.h
Comment (athanatos) [inline src/crimson/os/seastore/object_data_handler.h:118]: const

File: src/crimson/os/seastore/seastore.h
Comment (athanatos) [inline src/crimson/os/seastore/seastore.h:308]: const

File: src/common/options/crimson.yaml.in
Comment (athanatos) [inline src/common/options/crimson.yaml.in:68]: seastore_default_max_object_size

max_object_size is more meaningful to a user.

File: src/crimson/os/seastore/onode.h
Comment (athanatos) [inline src/crimson/os/seastore/onode.h:59]: const for all three of these.

**Review by athanatos — APPROVED** [review]
Other than const changes

**Comment by athanatos** [comment]
@xxhdx1985126 When I merge this, can I include your reviewed-by?

**Comment by xxhdx1985126** [comment]
@athanatos Done:-)

> @xxhdx1985126 When I merge this, can I include your reviewed-by?

Of course, my pleasure:-)

File: src/crimson/os/seastore/onode.h
Comment (cyx1231st) [inline src/crimson/os/seastore/onode.h:57]: Seems there is no user of this value.

File: src/crimson/os/seastore/object_data_handler.cc
Comment (cyx1231st) [inline src/crimson/os/seastore/object_data_handler.cc:271]: [max_object_size, &object_data]

**Review by cyx1231st — APPROVED** [review]
LGTM other than comments.

File: src/crimson/os/seastore/object_data_handler.cc
Comment (xxhdx1985126) [reply src/crimson/os/seastore/object_data_handler.cc:271]: max_block_size is not a local variable but a member of object data handler

File: src/crimson/os/seastore/onode.h
Comment (athanatos) [reply src/crimson/os/seastore/onode.h:57]: Yeah, looks like SeaStore passes it into ObjectDataHandler directly.

**Review by liu-chunmei — APPROVED** [review]

File: src/crimson/os/seastore/onode.h
Comment (athanatos) [reply src/crimson/os/seastore/onode.h:57]: I think we're just blocked on this one.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/onode.h]: @athanatos @cyx1231st Done:-)

**Comment by athanatos** [comment]
@xxhdx1985126 Needs a rebase, ping me after and I'll merge it.

**Comment by xxhdx1985126** [comment]
> @xxhdx1985126 Needs a rebase, ping me after and I'll merge it.

Done:-)

**Comment by athanatos** [comment]
@xxhdx1985126 Argh, forgot to add your reviewed-by after all, sorry about that!

**Comment by xxhdx1985126** [comment]
> @xxhdx1985126 Argh, forgot to add your reviewed-by after all, sorry about that!

No problem, don't worry:-)

**Pull request merged**
